### PR TITLE
Settle MSP promises on timeout, disconnect and CRC failure

### PR DESCRIPTION
**Pull request by blckmn**
`MSP.promise` never rejected: the per-request timer only logged and re-sent once, and `callbacks_cleanup` dropped pending callbacks without settling them on every tab switch and disconnect, so any outstanding `await MSP.promise(...)` hung forever — the stuck blackbox download spinner being the most visible casualty. Requests were also matched to responses by code alone, so two in-flight requests for the same code with different payloads could cross-resolve, and CRC-failed responses resolved "successfully" with a `crcError` flag almost nothing checked.

This ports the `_drain_cli_queue` discipline to the binary path:

- Promises reject with `MspTimeoutError` after retries (3 × 1s), `MspCancelledError` on queue drain or send-while-disconnected, and `MspCrcError` on checksum failure. Virtual mode still resolves `undefined`.
- Promise-based requests for the same code with different payloads are serialised, so responses can no longer cross-resolve. Plain `send_message` callbacks keep their existing behaviour throughout (never settled on timeout or drain, never parked); mixed promise/callback traffic for the same code behaves as before.
- `dataflashRead` moves onto the promise path and surfaces fatal errors to its callers, so an interrupted flash read now tears down cleanly instead of leaving the UI stuck.
- A global `unhandledrejection` guard swallows cancellation errors only — tab-switch drains of in-flight loads are routine, while timeout and CRC rejections stay visible as real signal. The hot paths (dataflash pull, onboard logging save, LED strip, connect helpers, OSD font upload, tuning sliders, STM32 reboot) handle rejections locally.

New coverage in `test/js/msp_promise.test.js`: timeout rejection and retry counts, disconnect and tab-switch drains, send-while-disconnected, virtual mode, CRC rejection, per-code serialisation and parked release, dedup, and legacy callback interplay.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved error handling across dataflash saving, dataflash reads, MSP-driven calculations, LED strip loading, serial initialization, and font uploads.
  * Prevented expected cancellations from showing as noisy unhandled promise rejections.
  * Strengthened MSP request reliability with retryable timeouts, improved deduping/queuing, and clearer CRC/timeout/cancel behavior.
* **New Features**
  * Added dedicated MSP error types (timeout, cancellation, CRC) to standardize error reporting.
* **Tests**
  * Added/expanded MSP promise semantics coverage for timeouts, disconnect/cleanup cancellation, CRC failures, and queued request behavior.

### CODE REVIEW

**Comment by haslinghuis** [comment]
Would it be possible to handle msp state (timeout / disconnect/ crc failures) and error handling in msp.js or does the consumer needs to handle it?

**Comment by nerdCopter** [comment]
AI Generated comment

Reviewed the diff. Timer cleanup, retry-then-reject, FIFO parking for 3+ same-code requests, and the `unhandledrejection` guard's `instanceof MspCancelledError` scoping all look correct — no double-settle or leak paths found.

One gap: per-code serialization only covers promise-vs-promise traffic. `process_data` still matches callbacks by `code` alone, so a legacy `send_message()` callback and an `MSP.promise()` call for the same code with different payloads can still cross-resolve. `test/js/msp_promise.test.js:1076-1096` ("does not park a promise behind an in-flight legacy request with a different payload") documents this directly, so it's a known limitation rather than an oversight — but most existing call sites are still legacy `send_message`, so the cross-resolution fix is scoped to promise-only traffic, not the whole codebase.

File: src/js/msp.js [hunk 1]
```diff
@@ -496,41 +499,56 @@ const MSP = {
...
         }
 
+        return this._transmit(code, data, callback_sent, callback_msp, false);
+    },
+    _buffer_matches(entry, byteLength, keyCrc) {
+        return (
+            entry.requestBuffer?.byteLength === byteLength &&
+            this.crc8_dvb_s2_data(new Uint8Array(entry.requestBuffer), 0, entry.requestBuffer.byteLength) === keyCrc
+        );
+    },
```
[the remark this answers is not in the document]
Comment (blckmn) [reply src/js/msp.js]: Fair point, a 1 in 256 false match at equal length would recreate exactly the cross-resolve this is meant to kill. Swapped to a byte for byte compare and dropped the CRC key entirely.

File: src/js/msp/MSPHelper.js [hunk 2]
```diff
@@ -2525,59 +2538,59 @@ MspHelper.prototype.dataflashRead = function (address, blockSize, onDataCallback
...
+        (error) => {
+            if (error instanceof MspCrcError) {
                 // Report crc error
                 console.log(`CRC error for address ${address} - retrying`);
                 onDataCallback(address, null); // returning null to the callback forces a retry
+            } else {
+                // Timeout or cancellation: surface the error as the fourth argument
+                onDataCallback(address, null, null, error);
             }
         },
```
[the remark this answers is not in the document]
Comment (blckmn) [reply src/js/msp/MSPHelper.js:2613]: Good catch, that was the same hang class this PR is chasing. Unknown compression types now error out through the fatal path rather than retrying, since a retry won't change what the firmware speaks.

File: src/components/tabs/OnboardLoggingTab.vue [hunk 3]
```diff
@@ -718,7 +718,18 @@ export default defineComponent({
...
+                            console.error("Error reading blackbox log:", error);
+                            gui_log(
+                                `<strong><span class="message-negative">${i18n.getMessage("error", {
+                                    errorMessage: error,
+                                })}</span></strong>`,
+                            );
+                            dismissSavingDialog();
+                            FileSystem.closeFile(openedFile);
+                            return;
+                        }
```
[the remark this answers is not in the document]
Comment (blckmn) [reply src/components/tabs/OnboardLoggingTab.vue:743]: Agreed, navigating away mid save shouldn't look like a failure. Cancellation now just dismisses the dialog and closes the file quietly, real errors keep the banner.

File: src/composables/ports/usePortsState.js [hunk 4]
```diff
@@ -88,9 +88,13 @@ export function usePortsState(getRules) {
...
-        MSP.promise(MSPCodes.MSP_VTX_CONFIG).then(() => {
-            mspHelper.loadSerialConfig(handleSerialConfigLoaded);
-        });
+        MSP.promise(MSPCodes.MSP_VTX_CONFIG)
+            .then(() => {
+                mspHelper.loadSerialConfig(handleSerialConfigLoaded);
+            })
+            .catch((error) => {
+                console.error("Failed to load VTX config for ports tab:", error);
+            });
```
[the remark this answers is not in the document]
Comment (blckmn) [reply src/composables/ports/usePortsState.js:101]: Done. The catch now clears the loading flag and still runs content_ready so the tab stays usable instead of spinning forever.

File: src/js/protocols/webstm32.js [hunk 5]
```diff
@@ -181,12 +182,24 @@ class STM32Protocol {
...
+                    disconnectFromMsp();
+                })
+                .catch((error) => {
+                    if (error instanceof MspCancelledError && error.reason === "disconnected") {
+                        // Expected: the FC drops the serial link as part of rebooting.
+                        disconnectFromMsp();
+                    } else {
+                        console.error(`${this.logHead} MSP_SET_REBOOT request failed:`, error);
+                    }
+                });
```
[the remark this answers is not in the document]
Comment (blckmn) [reply src/js/protocols/webstm32.js:203]: Added handleError() on the non cancellation path so connect_lock clears and the flashing state resets.

File: src/js/protocols/webstm32.js [hunk 6]
```diff
@@ -203,44 +216,48 @@ class STM32Protocol {
...
                 }
-            } catch (e) {
-                console.error(e);
-                this.reboot();
-            }
-        });
+            })
+            .catch((error) => {
+                console.error(`${this.logHead} MSP_BOARD_INFO request failed:`, error);
+            });
```
[the remark this answers is not in the document]
Comment (blckmn) [reply src/js/protocols/webstm32.js:262]: Same treatment as the reboot path, the catch now calls handleError() so the user can retry instead of being stuck.

**Comment by blckmn** [comment]
On the unsupported-reply point: not rejecting those. An unsupported reply is a legitimate negotiated outcome, older firmware answers optional probes that way, and rejecting would turn every one of those probes into an error on boards that are working fine. Resolved responses now carry an unsupported flag instead so callers that care can check it.

Both test suggestions taken: parked requests now have cleanup rejection coverage, and the cancellation assertions use instanceof.

**Review by haslinghuis — APPROVED** [review]
High-quality, well-reasoned fix for a real class of hangs. The parking/serialisation state machine holds up under the edge cases I traced, and it's backed by genuinely thorough tests.

**Verified on the PR branch:**
- New `test/js/msp_promise.test.js`: 14/14 pass
- Full suite: 404/404 pass (40 files)
- ESLint on all 12 changed files: clean

### What's good
- **Correct root cause.** The old `promise()` wrapped `send_message` with a resolve-only callback, so timeouts/CRC/drains never settled the promise → permanent hangs (the blackbox spinner). Porting the `_drain_cli_queue` discipline is the right model.
- **The parking state machine is robust.** I traced multi-park chains, deduped-sibling timeouts, and legacy/errorAware interleaving. `_release_parked` correctly guards on `callbacks.some(errorAware && code)`, so a parked request is never released while a deduped sibling is still in flight — no premature release, no leak. Every in-flight errorAware entry eventually calls `_release_parked` (response, timeout, or cleanup drains directly).
- **Legacy callbacks are genuinely untouched** on timeout/drain (verified by the "does not invoke legacy callback" tests). This backward-compat property is the crux, and it holds.
- Retry semantics match the description (`attempts` 1→3, three wire sends over 3s, reject at 3s).
- Tests encode the tricky invariants as comments (dedup only for payload-less requests, dispatch-by-code-alone) rather than hiding them.

### Considerations (non-blocking)

**1. Behavioural change for the ~200 unguarded `MSP.promise` call sites.** The PR handles the hot paths, but most tab `load()`/`save()` chains (`ReceiverTab`, `PidTuningTab`, `ConfigurationTab`, `SetupTab`, `ServosTab`, …) still do bare `await MSP.promise(...)`.
- `MspCancelledError` (tab-switch/disconnect) → swallowed by the global `unhandledrejection` guard. ✅ This converts old hangs into clean aborts.
- `MspTimeoutError` / `MspCrcError` → **not** swallowed. A genuine mid-load/mid-save timeout now aborts the chain and surfaces an `Uncaught (in promise)` in the console instead of hanging.

This is a net improvement (fail-fast vs hang-forever) and worth keeping. Just flagging that a timeout during e.g. a `save()` now aborts the remaining sequence and logs unhandled noise with no user-facing dialog on those non-hot paths. A follow-up could add a thin wrapper or a top-level timeout/CRC toast.

**2. `dataflashRead` unsupported-response path.** If the FC ever resolves `MSP_DATAFLASH_READ` with an empty/unsupported payload, `response.data.readU32()` in the `.then` throws → rejected promise with no local catch → unhandled rejection. This matches the old callback behaviour functionally (both log + stall the download), so no regression — just noting it survives the refactor.

**3. Minor.** `_on_timeout` resends deduped errorAware siblings independently, so two awaiters on the same buffer produce duplicate wire sends per retry. Rare-path efficiency nit only.

### Note on scope
Reviewed the branch as fetched from `pull/5249/head`. Worth a rebase check against current `master` before merge, since `msp.js`/`MSPHelper.js` are hot files.

Nice work — approving.

**Review by nerdCopter — APPROVED** [review]
